worker/src/main/java/gov/cms/ab2d/worker/service/JobProcessingServiceImpl.java

```diff
@@ -30,28 +59,167 @@ public Job putJobInProgress(String jobId) {
+             log.info("Sponsor {} is a parent sponsor. Processing children sponsors", sponsor.getOrgName());
+
+             return getContractsOfChildrenSponsor(sponsor);
+         } else {
+            return sponsor.getAttestedContracts();
+         }
+    }
+
+
+    private List<Contract> getContractsOfChildrenSponsor(Sponsor sponsor) {
+        final List<Sponsor> childrenSponsors = sponsorRepository.findByParent(sponsor);
+        return childrenSponsors.stream()
+                .map(s -> s.getAttestedContracts())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+
+    private List<JobOutput> processContract(final Path outputDir, Contract contract) {
+
+        var contractNumber = contract.getContractNumber();
+        var outputFile = fileService.createFile(outputDir, contractNumber + OUTPUT_FILE_SUFFIX);
+        var errorFile = fileService.createFile(outputDir, contractNumber + ERROR_FILE_SUFFIX);
+
+        var patientsByContract = beneficiaryAdapter.getPatientsByContract(contractNumber);
+        var patients = patientsByContract.getPatients();
+        int patientCount = patients.size();
+
+        // A mutex lock that all threads for a contract uses while writing into the shared files
+        var lock = new ReentrantLock();
+        var futureResourcesHandles = patients.stream()
+                .map(patient -> patientClaimsProcessor.process(patient.getPatientId(), lock, outputFile, errorFile))
+                .collect(Collectors.toList());
+
+        int errorCount = processHandles(futureResourcesHandles);
+        while (!futureResourcesHandles.isEmpty()) {
+            sleep();
+            errorCount += processHandles(futureResourcesHandles);
+        }
+
+        final List<JobOutput> jobOutputs = new ArrayList<>();
+        if (errorCount < patientCount) {
+            final JobOutput jobOutput = createPartialJobOutput(outputFile);
+            jobOutput.setError(false);
+            jobOutputs.add(jobOutput);
+        }
+        if (errorCount > 0) {
+            final JobOutput jobOutput = createPartialJobOutput(errorFile);
+            jobOutput.setError(true);
+            jobOutputs.add(jobOutput);
+        }
+
+        return jobOutputs;
+    }
+
+    private int processHandles(List<Future<Integer>> futureResourcesHandles) {
+        int errorCount = 0;
+
+        var iterator = futureResourcesHandles.iterator();
+        while (iterator.hasNext()) {
+            var future = iterator.next();
+            if (future.isDone()) {
+                try {
+                    var responseCount = future.get();
+                    if (responseCount > 0) {
+                        errorCount += responseCount;
+                    }
+                } catch (InterruptedException  e) {
+                    ++errorCount;
+                    log.error("interrupted exception while processing patient ", e);
+                } catch (ExecutionException e) {
```

Comment:
This exception means that the Future ended with an uncaught exception, which was likely a non-recoverable condition. Let's re-throw it as RuntimeException and let the entire Job to fail.

Reply:
This comment is only for the `ExecutionException` ? 
Or should I do the same for `InterruptedException` as well?

Reply:
Good question. Let's do both.

Reply:
done